pinc/ProofProject.inc

```diff
@@ -0,0 +1,217 @@
+        }
+        return $this->render_page_data();
+    }
+
+    public function render_page_text(): array
+    {
+        // validate user and page? page should be in out or temp state only
+        return $this->get_page_text_array();
+    }
+
+    private function render_page_data(): array
+    {
+        $round_info_array = [];
+        foreach ($this->round->other_rounds_with_visible_usernames as $other_round_id) {
+            $other_round = get_Round_for_round_id($other_round_id);
+            $username = $this->get_username_for_round($other_round);
+            $round_info_array[] = [$other_round->id, $username];
+        }
+        return
+            $this->get_page_text_array() +
+            $this->get_page_state_array() +
+            [
+                "pagenum" => pathinfo($this->page_name, PATHINFO_FILENAME),
+                "round_info_array" => $round_info_array,
+                "pagename" => $this->page_name,
+                "image_url" => "{$this->project->url}/{$this->page_name}",
+                "langDir" => lang_direction(langcode2_for_langname($this->project->languages[0])),
+            ];
+    }
+
+    public function get_page_text_array(): array
+    {
+        return ["text" => $this->get_text()];
+    }
+
+    private function get_page_state_array()
+    {
+        return [
+            "pagestate" => $this->page_state,
+            "saved" => $this->can_be_reverted_to_last_save(),
+        ];
+    }
+
+    private function get_valid_text(string $text): string
+    {
+        $pattern_string = build_character_regex_filter($this->project->get_valid_codepoints());
+        foreach (split_graphemes($text) as $grapheme) {
+            if (1 != preg_match("/$pattern_string/u", $grapheme)) {
+                throw new ProjectInvalidTextException(_("The text contains characters which are not allowed for this project"));
+            }
+        }
+        return _normalize_page_text($text, $this->project->projectid);
+    }
+}
+
+class ProofProject
+{
+    public function __construct(object $project)
+    {
+        global $pguser;
+
+        $project_round = $project->get_project_available_round();
+        $project_round->validate_user_can_access($pguser);
+
+        $this->project = $project;
+        $this->round = $project_round;
+    }
+
+    public function checkout(): array
+    {
+        global $pguser;
+
+        [$page_name, $state] = get_available_proof_page_array($this->project, $this->round, $pguser);
+        $project_page = new ProjectPage($page_name, $state);
+        $proof_project_page = new ProofProjectPage($this, $project_page);
+
+        return $proof_project_page->checkout($pguser);
+    }
+}
+
+class TextChecker
+{
+    private const VALID_TEXT = 0;
+    private const INVALID_TEXT = 1;
+
+    private static $text_array;
+    private static $valid_text_block;
+
+    public static function analyse(string $page_text, object $project): array
```

Comment:
Consider making this a non-static class. It feels odd that this class is collecting & persisting data statically.

Reply:
Yes I can do that.